pkg/store/fsm.go

```diff
@@ -253,69 +274,101 @@ func (f fsmSnapshot) Release() {
-			// empty case, e.g. just created namespace
-			enforcers[key.(string)] = EnforcerState{}
-		}
-		return true
-	})
 	var err error
 	fsm := &fsmSnapshot{
 		startT: time.Now(),
 		logger: s.logger,
 	}
-	fsm.enforcers, err = json.Marshal(enforcers)
+	writer := new(bytes.Buffer)
+	err = s.enforcersState.Snapshot(writer)
 	if err != nil {
-		s.logger.Printf("failed to encode Enforcers for snapshot: %s", err.Error())
+		s.logger.Printf("failed to encode enforcerState: %s", err.Error())
 		return nil, err
 	}
+	models := make(map[string]string)
+	s.enforcers.Range(func(key, value interface{}) bool {
+		if e, ok := value.(*casbin.DistributedEnforcer); ok {
+			models[key.(string)] = e.GetModel().ToText()
+		}
+		return true
+	})
+	fsm.state = writer.Bytes()
 	fsm.meta, err = json.Marshal(s.meta)
 	if err != nil {
-		s.logger.Printf("failed to encode Meta for snapshot: %s", err.Error())
+		s.logger.Printf("failed to encode Meta: %s", err.Error())
+		return nil, err
+	}
+	fsm.models, err = json.Marshal(models)
+	if err != nil {
+		s.logger.Printf("failed to encode Meta: %s", err.Error())
 		return nil, err
 	}
 	return fsm, nil
 }
 
 // Restore restores form a preexisted states
 func (s *Store) Restore(closer io.ReadCloser) error {
+	var err error
 	var data persistData
-	err := json.NewDecoder(closer).Decode(&data)
+	err = json.NewDecoder(closer).Decode(&data)
 	if err != nil {
+		s.logger.Println("failed to decode restore data", err)
 		return err
 	}
-	var meta map[string]map[string]string
-	err = json.Unmarshal(data.Meta, &meta)
+
+	err = s.enforcersState.Restore(bytes.NewReader(data.State))
 	if err != nil {
+		s.logger.Println("failed to restore enforcer state", err)
 		return err
 	}
-	var enforcers map[string]EnforcerState
-	err = json.Unmarshal(data.Enforcers, &enforcers)
+	s.enforcers = sync.Map{}
+	models := make(map[string]string)
+	err = json.Unmarshal(data.Models, &models)
 	if err != nil {
+		s.logger.Println("failed to unmarshal models state", err)
 		return err
 	}
-	s.enforcers = sync.Map{}
-	for k, v := range enforcers {
-		e, err := casbin.NewDistributedEnforcer()
-		if err != nil {
-			return err
-		}
-		m, err := CreateModelFormEnforcerState(v)
-		if err != nil {
-			return err
+	err = s.enforcersState.Foreach(func(name []byte, b *bolt.Bucket) error {
+		if model, ok := models[string(name)]; ok {
+			enforcer, err := casbin.NewDistributedEnforcer()
+			if err != nil {
+				s.logger.Println("failed to create enforcer", err)
+				return err
+			}
+			a, err := adapter.NewAdapter(s.enforcersState, string(name), "")
+			if err != nil {
+				s.logger.Println("failed to create adapter", err)
+				return err
+			}
+			model, err := model2.NewModelFromString(model)
+			if err != nil {
+				s.logger.Println("failed to create model", err)
+				return err
+			}
+			err = a.LoadPolicy(model)
+			if err != nil {
+				s.logger.Println("failed to load policy", err)
+				return err
+			}
+			err = enforcer.InitWithModelAndAdapter(model, a)
```

Comment:
`InitWithModelAndAdapter` includes `LoadPolicy` and `BuildRoleLinks`, so don't call that.